### PR TITLE
OOP school library: unit tests

### PR DESCRIPTION
**Pull request by RitobrotoMukherjee**
### Project requirements that I have performed:
- **Used `AAA` pattern**
- Used proper `Mocks` for dependent methods
- Added unit tests to classes:
  - `Book`
  - `Person`
  - `Student`
  - `Teacher`
  - `Classroom`
  - `Rental`
  - all classes that I created while implementing the **Decorator design pattern**
    - `BaseDecorator`
    - `CapitalizeDecorator`
    - `TrimmerDecorator`

## Note for the CR:
- **I have been added to a new group as my old partner is not available as he stays in the currently flooded area of Pakistan. The new group of people had already finished this project, So, I had to do the work alone for my project**
CR

### CODE REVIEW

**Comment by bellom** [comment]
> * I have been added to a new group as my old partner is not available as he stays in the currently flooded area of Pakistan. The new group of people had already finished this project, So, I had to do the work alone for my project

@RitobrotoMukherjee - For any issue that was resolved by Student Success, please always attach a screenshot so that the CR can view it and review your PR without a delay.

**Comment by RitobrotoMukherjee** [comment]
> > * I have been added to a new group as my old partner is not available as he stays in the currently flooded area of Pakistan. The new group of people had already finished this project, So, I had to do the work alone for my project
> 
> @RitobrotoMukherjee - For any issue that was resolved by Student Success, please always attach a screenshot so that the CR can view it and review your PR without a delay.

@bellom What should I add here, can you please suggest? Cause there is around 6-7 mails in a mail trail, how can I add a screenshot for that, bit confused on this point!!

Also in the current group where I have been added to had already finished this project and got it approved, before I got added to the group. So, in this project I had to work alone.

**FYI: The last and next project for this week I had worked with my current team and is under approval now.**

**Comment by bellom** [comment]
@RitobrotoMukherjee - You can simply add only the email related to you joining the new group from Microverse. I believe you reported the issue with your partner and you were added to the new group.

I will review this PR manually, hang in there....

**Comment by RitobrotoMukherjee** [comment]
> @RitobrotoMukherjee - You can simply add only the email related to you joining the new group from Microverse. I believe you reported the issue with your partner and you were added to the new group.
> 
> I will review this PR manually, hang in there....

@bellom Sure, I shall be waiting. And I will add the screenshot of that mail subject with the final reply in the PR right away, thank you for the guidance and help.

File: main.rb [hunk 1]
```diff
@@ -1,5 +1,4 @@
 # rubocop:disable Metrics/CyclomaticComplexity
-# rubocop:disable Metrics/PerceivedComplexity
 # rubocop:disable Metrics/MethodLength
```
Comment (bellom) [inline main.rb:2]: ## [Optional]

If you follow the single responsibility principle, you wouldn't have disabled this linter config 🙂 - kindly read more [here](https://thoughtbot.com/blog/sandi-metz-rules-for-developers)

**Review by bellom — APPROVED** [review]
@RitobrotoMukherjee

### This is a manual review|

Note: It will be approved on your dashboard as soon as possible. 

# Approved

Cheers and Happy coding!👏👏👏

File: main.rb (hunk 1 above)
Comment (RitobrotoMukherjee) [reply main.rb:2]: @bellom I didn't find any valid real-world reason in the context of this project in the provided link for the main menu options. In real-world applications, this data would usually come as a JSON or YAML object which could be directly mapped using `loops` without the `switch cases`. In that case, method lengths will automatically reduce, but here the main has to take inputs and pass them forward.

**N.B: Now how can that be shortened for different `cases` I have no valid idea. If you could guide me that would be great. I can meet with you over a zoom**

**In my advocate:** When you check the `main` function, it's having only one responsibility as project required - 
 > taking user inputs & passing to a dependent object (`app` here)

- It's open and closed, just adding an extra 1 case will take care of another functionality
- The Inherited classes parent and child classes are substitutable (**Liskov substitution**)
- Similar types of classes are having different responsibilities but similar implementations using interfaces - (**Interface segregation**)
- In the models, the **Dependency Inversion** followed while using **decorator pattern**

**Comment by RitobrotoMukherjee** [comment]
Hi @bellom , Thank you for checking and approving my project.🤝

**Pull request merged**